### PR TITLE
Grunt

### PR DESCRIPTION
**Pull request by AndreTheHunter**
The big grunt update!
* tests are now far easier to to run by using grunt
* split test/runner.js into setup.js, helper.js and various config in the Gruntfile.js

See the commit logs for more information

Some further work required:
* get `grunt test-mobile` to work
* istanbul code coverage
* Need consistent assertion style. I vote for 'expect' bdd style
* 'use strict' on all modules
* fix ErrorHandler.RuntimeError causing process exit

### CODE REVIEW

**Comment by AndreTheHunter** [comment]
I've added appium, but getting a `Requested a new session but one was in progress` error.
Run `grunt test-mobile --spect=test/spec/mobile/context.js`

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                grunt.config.set(testSrc, ['test/spec/' + specDir[0] + '/**/*.js']);
+            } else {
+                // otherwise test global + device specific test specs
+                grunt.config.set(testSrc, ['test/spec/' + env + '/**/*.js', 'test/spec/*.js']);
+            }
+        }
+    }
+
+    //TODO istanbul code coverage
+    // "coverage": "./node_modules/.bin/istanbul cover -x \"**/helpers/_*.js\" ./test/runner.js",
```
Comment (christian-bromann) [inline Gruntfile.js]: What about this?

Comment (AndreTheHunter) [inline Gruntfile.js]: I'm sorry I don't understand. What about it?
That code was originally from test/runner.js, if it needs to change let me know what the plan is.

Comment (christian-bromann) [inline Gruntfile.js]: mhm .. I can't find it in `/test/runner.js`. We need to make sure that istanbul get triggered if we run our tests in travis

**Comment by AndreTheHunter** [comment]
So what will it take to get this merged?
I'm not in love with using grunt, but rather the fact it makes running tests a lot easier and that npm installs all the required programs. If it's easier for people to run the tests, more people will write tests and make it easier to contribute.
If you have any objections to using grunt, please discuss it here (rather than letting this PR just stagnate)

File: package.json [hunk 2]
```diff
@@ -16,23 +16,19 @@
...
-    "test-mobile": "_APP=safari _BROWSER=safari _PLATFORMVERSION=7.1 _APPIUMVERSION=1.2 _PLATFORMNAME=iOS _DEVICENAME=iPhone_Simulator _PORT=4723 _ENV=mobile node ./test/runner.js",
-    "test-desktop": "_ENV=desktop _BROWSER=chrome node ./test/runner.js",
-    "test-functional": "_ENV=functional _BROWSER=phantomjs node ./test/runner.js",
-    "test-multibrowser": "_ENV=multibrowser node ./test/runner.js",
+    "test": "grunt test-all",
+    "test-mobile": "grunt test-mobile",
+    "test-desktop": "grunt test-desktop",
+    "test-functional": "grunt test-functional",
+    "test-multibrowser": "grunt test-multibrowser",
     "coverage": "./node_modules/.bin/istanbul cover -x \"**/helpers/_*.js\" ./test/runner.js",
```
Comment (christian-bromann) [inline package.json]: I guess `./test/runner.js` is not that what we want to execute here

**Comment by christian-bromann** [comment]
@AndreTheHunter one concern is that code coverage is not working with this setup (see my comments above). Also do you have travis hooked up with your fork so I can check if it's working properly? I don't care about test results, only that all tests get executed.

Sorry if you get the feeling that I am absent but I am working on the `v3` docs. I don't think it is a good idea to merge this PR before releasing `v3`. Therefor could you create a new PR pointing to master as I've merged everything yesterday. Sorry for the inconvenience.

Thanks for all your work!

File: package.json (hunk 2 above)
Comment (AndreTheHunter) [inline package.json]: Correct. test/runner.js has been split into test/setup.js, tes/helper.js and various config in the Gruntfile.js so now all the test targets are run via grunt. See commit logs [c67d0b2](https://github.com/AndreTheHunter/webdriverio/commit/c67d0b2249b101df7794e600916110a5118a9614).
A nice feature of npm is we can run the grunt-cli without needing to have grunt installed globally! This makes it easier to run the tests, people/CI just have to run `npm test-functional`.

File: Gruntfile.js (hunk 1 above)
Comment (AndreTheHunter) [inline Gruntfile.js]: Istanbul, that is something I am currently working on. I'll add that to the new PR.

**Comment by AndreTheHunter** [comment]
@christian-bromann Once istanbul is working under grunt, I'll get it running under travis.
I'll create a new PR shortly.
I understand you must be busy, thanks for your time thus far.

**Pull request closed without merging**